src/block/block.service.ts

```diff
@@ -0,0 +1,133 @@
+import { providers } from "ethers";
+import { Block } from "src/models/Block";
+import { OperationFactory } from "src/models/OperationFactory";
+import { PartialBlockIdentifier } from "src/models/PartialBlockIdentifier";
+import { Transaction } from "src/models/Transaction";
+import { TransactionIdentifier } from "src/models/TransactionIdentifer";
+import { getRPCProvider } from "src/utils/client";
+
+import { RewardService } from "./reward.service";
+
+@Injectable()
+export class BlockService {
+  constructor(private rewardService: RewardService) {}
+
+  public async getBlock(block?: number) {
+    const provider = getRPCProvider();
+    const {
+      transactions,
+      miner,
+      timestamp,
+      parentHash,
+      hash,
+      number: blockNumber,
+    } = await provider.getBlockWithTransactions(block || "latest");
+
+    const parent = await provider.getBlock(parentHash);
+
+    const receipts = await Promise.all(
+      transactions.map((tx) => provider.getTransactionReceipt(tx.hash))
+    );
+
+    const operationFactory = new OperationFactory();
+
+    const blockTransaction = await this.buildBlockTransactions(
+      miner,
+      receipts,
+      transactions,
+      operationFactory
+    );
+
+    const rewardTransaction = await this.getRewardTransaction(
+      blockNumber,
+      miner,
+      operationFactory
+    );
+
+    return {
+      block: new Block(
+        new PartialBlockIdentifier(blockNumber, hash),
+        new PartialBlockIdentifier(parent.number, parent.hash),
+        timestamp,
+        [...blockTransaction, rewardTransaction]
+      ),
+    };
+  }
+
+  public async getBlockTransaction(
+    blockNumber: number,
+    transactionHash: string
+  ) {
+    const provider = getRPCProvider();
+
+    const transaction = await provider.getTransaction(transactionHash);
+
+    if (transaction.blockNumber !== blockNumber) {
+      throw new Error("Blocknumber mismatch");
+    }
+
+    const receipt = await transaction.wait();
+    const block = await provider.getBlock(blockNumber);
+
+    const blockTransaction = await this.buildBlockTransactions(
+      block.miner,
+      [receipt],
+      [transaction],
+      new OperationFactory()
+    );
+
+    return {
+      transaction: blockTransaction,
+    };
+  }
+
+  private async getRewardTransaction(
+    blockNumber: number,
+    miner: string,
+    operationFactory: OperationFactory
+  ) {
+    const blockRewards = await this.rewardService.calculateBlockRewards(
+      miner,
+      blockNumber
+    );
+
+    return new Transaction(
+      null,
+      blockRewards.flatMap(({ address, value }) =>
+        operationFactory.reward(address, value)
+      )
+    );
+  }
+
+  private async buildBlockTransactions(
+    miner: string,
+    receipts: providers.TransactionReceipt[],
+    transactions: providers.TransactionResponse[],
+    operationFactory: OperationFactory
+  ) {
+    const transactionCache = transactions
+      .filter((tx) => !tx.value.isZero())
```

Comment:
it should be tested if this also yields the transactions sent from a smart-contract (inner transactions). This can be tested by sending EWT to a smart contract and then having the smart contract send those EWT to another address in a subsequent transaction

Reply:
I think I will remove this check and return all transactions from the block, 0 or non zero valued.

Reply:
I assumed that due to the description from Rosetta API:

_Block endpoints are used to access any data stored in a block. It is critical that these Endpoints can be used to fetch all balance-changing operations in a block. If this is not possible, clients that reconcile balances on accounts they care about will not be able to use the implemented interface. Balance reconciliation, in this scenario, means comparing what balance the node returns for an account with the balance that can be calculate from looking at all the account's transactions._

Especially the part of _It is critical that these Endpoints can be used to fetch all balance-changing operations in a block_